adc/ade7913/ade7913.c

```diff
@@ -0,0 +1,304 @@
+#define ADE7913_STATUS0_IC_PROT  (1 << 2)
+
+#define ADE7913_SYNC_SNAP_SYNC (1 << 0)
+#define ADE7913_SYNC_SNAP_SNAP (1 << 1)
+
+#define ADE7913_EMI_CTRLS(X) (1 << X)
+
+#define ADE7913_STATUS1_VERSION (1 << 0)
+#define ADE7913_STATUS1_ADC_NA  (1 << 3)
+
+/* ADE7913 communication utils */
+#define ADE7913_READ_BIT         0x4
+#define ADE7913_ADDR_OFFS        0x3
+#define ADE7913_LOCK_SEQ         0xca
+#define ADE7913_UNLOCK_SEQ       0x9c
+#define ADE7913_CONFIG_FREQ_8KHZ (0x00 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_CONFIG_FREQ_4KHZ (0x01 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_CONFIG_FREQ_2KHZ (0x10 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_CONFIG_FREQ_1KHZ (0x11 << ADE7913_CONFIG_ADC_FREQ)
+#define ADE7913_READY            1
+
+
+static int lpspi_transaction(oid_t *device, int cs,
+	uint8_t *buff, size_t size)
+{
+	msg_t msg;
+	multi_i_t *imsg = (multi_i_t *)msg.i.raw;
+	multi_o_t *omsg = (multi_o_t *)msg.o.raw;
+
+	msg.type = mtDevCtl;
+	msg.i.data = buff;
+	msg.i.size = size;
+	msg.o.data = buff;
+	msg.o.size = size;
+
+	imsg->id = device->id;
+	imsg->spi.type = spi_transaction;
+	imsg->spi.transaction.cs = cs;
+	imsg->spi.transaction.frameSize = size;
+
+	msgSend(device->port, &msg);
+
+	return omsg->err;
+}
+
+
+static int ade7913_read_reg(oid_t *device, int cs,
+	uint8_t addr, uint8_t *data, size_t size)
+{
+	uint8_t buff[size + 1];
+
+	if (size == 0)
+		return EOK;
+
+	if (addr > ADE7913_TEMPOS)
+		return -EINVAL;
+
+	buff[0] = (addr << ADE7913_ADDR_OFFS) | ADE7913_READ_BIT;
+
+	if (lpspi_transaction(device, cs, buff, size) < 0)
+		return -EIO;
+
+	memcpy(data, &buff[1], size);
+
+	return EOK;
+}
+
+
+/* Only 1 byte write at a time possible */
+static int ade7913_write_reg(oid_t *device, int cs,
+	uint8_t addr, uint8_t value)
+{
+	int res = 0;
+	uint8_t buff[2] = { (addr << ADE7913_ADDR_OFFS), value };
+
+	/* Check if register is writtable */
+	if ((addr >= ADE7913_IWV && addr <= ADE7913_CNT_SNAPSHOT) ||
+		addr > ADE7913_EMI_CTRL || addr == ADE7913_STATUS0)
+		return -EINVAL;
+
+	if ((res = lpspi_transaction(device, cs, buff, 2)) < 0)
+		return res;
+
+	if (addr == ADE7913_LOCK)
+		return EOK;
+
+	/* Check written register */
+	buff[0] = (addr << ADE7913_ADDR_OFFS) | ADE7913_READ_BIT;
+
+	if ((res = lpspi_transaction(device, cs, buff, 2)) < 0)
+		return res;
+
+	return value == buff[1] ? EOK : -EAGAIN;
+}
+
+
+static int ade7913_set_clear_bits(oid_t *device, int cs,
+	uint8_t addr, uint8_t set, uint8_t clear)
+{
+	int res;
+	uint8_t val;
+
+	if ((res = ade7913_read_reg(device, cs, addr, &val, sizeof(val))) < 0)
+		return res;
+
+	val |= set;
+	val &= ~clear;
+
+	if ((res = ade7913_write_reg(device, cs, addr, val)) < 0)
+		return res;
+
+	return EOK;
+}
+
+
+int ade7913_ready(oid_t *device, int cs)
+{
+	uint8_t reg;
+
+	if (ade7913_read_reg(device, cs, ADE7913_STATUS0, &reg, sizeof(reg)))
+		return -EIO;
+
+	if (!(reg & ADE7913_STATUS0_RESET_ON))
+		return ADE7913_READY;
+
+	return !ADE7913_READY;
```

Comment:
This define `ADE7913_READY` is really no needed. It's only used here to return either zero or one.

Reply:
Removed define